### PR TITLE
#59 - Added implementation for InMemoryStorage.exists()

### PR DESCRIPTION
**Pull request by olegmoz**
Part of #59
Added exists() method and tests for it

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#72](https://github.com/artipie/asto/issues/72) is assigned to @fanifieiev[/z](https://www.0crat.com/u/fanifieiev), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/9b9da5cd-025d-4ade-9869-e6c8d6ea485a); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by g4s8** [comment]
@0crat assign @amihaiemil

**Comment by 0crat** [comment]
> @0crat assign @amihaiemil  ([here](https://github.com/artipie/asto/issues/72#issuecomment-588060935))

@g4s8 This pull request [#72](https://github.com/artipie/asto/issues/72) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/8ee45b39-7898-403c-8a5a-29d6a3f40aae); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by g4s8** [comment]
@0crat refuse

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/artipie/asto/issues/72#issuecomment-589562012))

@g4s8 The user @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) resigned from [#72](https://github.com/artipie/asto/issues/72), please stop working. Reason for job resignation: Order was cancelled

**Comment by 0crat** [comment]
This pull request [#72](https://github.com/artipie/asto/issues/72) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/51925410-004c-4a14-9274-34c0874548c0); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/test/java/com/artipie/asto/InMemoryStorageTest.java [hunk 1]
```diff
@@ -40,6 +40,32 @@
  */
 public final class InMemoryStorageTest {
 
+    @Test
+    void shouldExistForSavedKey() throws Exception {
+        final Storage storage = this.storage();
+        final Key key = new Key.From("shouldExistForSavedKey");
+        final byte[] data = "some data".getBytes();
+        new BlockingStorage(storage).save(key, data);
+        MatcherAssert.assertThat(storage.exists(key).get(), Matchers.equalTo(true));
```
Comment (paulodamaso) [inline src/test/java/com/artipie/asto/InMemoryStorageTest.java]: @olegmoz Please add an assertion failure message and do not use static matchers

File: src/test/java/com/artipie/asto/InMemoryStorageTest.java [hunk 2]
```diff
@@ -40,6 +40,32 @@
...
+    }
+
+    @Test
+    void shouldNotExistForUnknownKey() throws Exception {
+        final Storage storage = this.storage();
+        final Key key = new Key.From("shouldNotExistForUnknownKey");
+        MatcherAssert.assertThat(storage.exists(key).get(), Matchers.equalTo(false));
```
Comment (paulodamaso) [inline src/test/java/com/artipie/asto/InMemoryStorageTest.java]: @olegmoz Here also

File: src/test/java/com/artipie/asto/InMemoryStorageTest.java [hunk 3]
```diff
@@ -40,6 +40,32 @@
...
+    }
+
+    @Test
+    void shouldNotExistForParentOfSavedKey() throws Exception {
+        final Storage storage = this.storage();
+        final Key parent = new Key.From("shouldNotExistForParentOfSavedKey");
+        final Key key = new Key.From(parent, "child");
+        final byte[] data = "content".getBytes();
+        new BlockingStorage(storage).save(key, data);
+        MatcherAssert.assertThat(storage.exists(parent).get(), Matchers.equalTo(false));
```
Comment (paulodamaso) [inline src/test/java/com/artipie/asto/InMemoryStorageTest.java]: @olegmoz Here also

**Comment by olegmoz** [comment]
@paulodamaso all fixed, please review

File: src/test/java/com/artipie/asto/InMemoryStorageTest.java [hunk 4]
```diff
@@ -39,6 +40,44 @@
...
+            new IsEqual<>(true)
+        );
+    }
+
+    @Test
+    void shouldNotExistForUnknownKey() throws Exception {
+        final Storage storage = this.storage();
+        final Key key = new Key.From("shouldNotExistForUnknownKey");
+        MatcherAssert.assertThat(
+            "Key that was never saved should nto exist",
```
Comment (paulodamaso) [inline src/test/java/com/artipie/asto/InMemoryStorageTest.java]: @olegmoz PLease fix the typo here

Comment (olegmoz) [reply src/test/java/com/artipie/asto/InMemoryStorageTest.java]: @paulodamaso thank you! Fixed

**Comment by olegmoz** [comment]
@paulodamaso please review

**Comment by olegmoz** [comment]
@g4s8 fixed the PR after master merge, ready to be merged into master

**Comment by paulodamaso** [comment]
@olegmoz Great, thanks

**Comment by paulodamaso** [comment]
@g4s8 please merge

**Comment by 0crat** [comment]
Order was finished: +15 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Pull request merged**